### PR TITLE
Skip redirection to approval when it is not required (#2686)

### LINKED ISSUE #2686: Skip redirection to approval when it's not required
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/dexidp/dex/blob/master/.github/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/dexidp/dex/issues) for an issue that matches the one I want to file, without success.

### Problem Description

There is an unnecessary redirection to /approval when in fact it is not required by configuration.

### Proposed Solution

Skip the redirection and send the code directly back to the client.

### Alternatives Considered

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by nobuyo**
#### Overview

Fixes #2686.

#### What this PR does / why we need it

Skip redirection to `/approval` if it is not required.

#### Special notes for your reviewer

I've locally made `handleApproval` always return an error.
And I ran tests and confirmed that it does not fail (= `/approval` is not accessed).

#### Does this PR introduce a user-facing change?

```release-note
enhance: Skip redirection to `/approval` when it's not required
```

### CODE REVIEW

File: server/handlers.go [hunk 1]
```diff
@@ -460,19 +471,30 @@ func (s *Server) handleConnectorCallback(w http.ResponseWriter, r *http.Request)
...
+		return
+	}
+
 	http.Redirect(w, r, redirectURL, http.StatusSeeOther)
 }
 
 // finalizeLogin associates the user's identity with the current AuthRequest, then returns
 // the approval page's path.
-func (s *Server) finalizeLogin(identity connector.Identity, authReq storage.AuthRequest, conn connector.Connector) (string, error) {
+func (s *Server) finalizeLogin(identity connector.Identity, authReq storage.AuthRequest, conn connector.Connector) (returnURL string, canSkipApproval bool, err error) {
```
Comment (sagikazarmark) [inline server/handlers.go:497]:
```suggestion
func (s *Server) finalizeLogin(identity connector.Identity, authReq storage.AuthRequest, conn connector.Connector) (string, bool, error) {
```

File: server/handlers.go [hunk 2]
```diff
@@ -488,8 +510,8 @@ func (s *Server) finalizeLogin(identity connector.Identity, authReq storage.Auth
 		a.ConnectorData = identity.ConnectorData
 		return a, nil
 	}
-	if err := s.storage.UpdateAuthRequest(authReq.ID, updater); err != nil {
```
Comment (sagikazarmark) [inline server/handlers.go:491]: Why remove the shadowing?

**Review by sagikazarmark — CHANGES_REQUESTED** [review]
Thank you for your PR!

As far as I can tell some of the changes are rather the result of personal taste. Please revert those, unless absolutely necessary and related to this change.

Also, please provide tests verifying this new behavior.

Thanks!

File: server/handlers.go (hunk 2 above)
Comment (nobuyo) [reply server/handlers.go:491]: Sorry my mistake. I'll revert it too.

File: server/handlers.go (hunk 1 above)
Comment (nobuyo) [reply server/handlers.go]: I've reverted those changes. Apologize for ignoring your suggestions (there were several other places needed to be revert).

**Comment by nobuyo** [comment]
Thank you for your review! I've updated the code and added tests.

**Comment by nobuyo** [comment]
BTW, the CodeQL code scanning is failing, but error appears to be in a place that I have not directly changed 🤔 
Is there anything I have to do?

**Comment by nobuyo** [comment]
@sagikazarmark @nabokihms Could you please take a look at this? Or does this issue no longer block next release?

**Comment by nabokihms** [comment]
@nobuyo, everything seems good to me. Thank you for contributing. I will approve the PR. 

Let's wait for @sagikazarmark to say his final word since he was the initial reviewer.

**Review by nabokihms — APPROVED** [review]

**Review by sagikazarmark — APPROVED** [review]
LGTM, thanks @nobuyo

**Pull request merged**